### DISCUSSION TITLE
Can I define Analysis Tools within the configuration?

### DISCUSSION THREAD

**Discussion by stevenjh** [TerriaJS/terriajs#5499] [Q&A]
I'm wanting to run some analysis on a spatial area defined by the user, doing some server side processing to produce a chart (not in terria at this stage) and return the data, opening a new window.

I'm reviewing the Analysis Tools at https://map.geo-rapp.org/ and looking at how I can leverage the same UI/approach, but without  having a WPS Server setup.

The approach I'm considering is:

1. define in config as below, pointing at terriajs-server
```
{
          "name": "wps_emulate_test",
          "type": "wps",
          "url": "http://localhost:3001/functions",
          "identifier": "area_test",
          "description": "testing"
}
```
2. emulate the wps xml response as below
```
<wps:ProcessDescriptions xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsDescribeProcess_response.xsd" service="WPS" version="1.0.0" xml:lang="en-US">
<ProcessDescription wps:processVersion="2" storeSupported="true" statusSupported="true">
	<ows:Identifier>geometryDrill</ows:Identifier>
	<ows:Title>Monthly Geometry Drill (Means)</ows:Title>
	<ows:Abstract></ows:Abstract>
	<ows:Metadata xlink:title="TimeSeries Extractor"/>
	<DataInputs>
		
		<Input minOccurs="0" maxOccurs="1">
			<ows:Identifier>geometry_id</ows:Identifier>
			<ows:Title>Point/Polygon name</ows:Title>
			<ows:Abstract></ows:Abstract>
			<LiteralData>
				<ows:DataType ows:reference=""></ows:DataType>
				
				<ows:AnyValue />
				
			</LiteralData>
		</Input>
		
		
		<Input minOccurs="1" maxOccurs="1">
			<ows:Identifier>geometry</ows:Identifier>
			<ows:Title>Geometry</ows:Title>
			<ows:Abstract></ows:Abstract>
			<ComplexData>
				<Default>
					<Format>
						<MimeType>application/vnd.geo+json</MimeType>
						<Schema>http://geojson.org/geojson-spec.html#geojson</Schema>
					</Format>
				</Default>
				<Supported>
					<Format>
						<MimeType>application/vnd.geo+json</MimeType>
						<Schema>http://geojson.org/geojson-spec.html#geojson</Schema>
					</Format>
				</Supported>
			</ComplexData>
		</Input>
		
	</DataInputs>
	<ProcessOutputs>
		<Output>
			<ows:Identifier>Result</ows:Identifier>
			<ows:Title>Time Series Output</ows:Title>
			<ows:Abstract>Time series data for location.</ows:Abstract>
			<ComplexOutput>
				<Default>
					<Format>
						<MimeType>application/vnd.terriajs.catalog-member+json</MimeType>
						<Schema>https://tools.ietf.org/html/rfc7159</Schema>
					</Format>
				</Default>
				<Supported>
					<Format>
						<MimeType>application/vnd.terriajs.catalog-member+json</MimeType>
						<Schema>https://tools.ietf.org/html/rfc7159</Schema>
					</Format>
				</Supported>
			</ComplexOutput>
		</Output>
	</ProcessOutputs>
</ProcessDescription>
</wps:ProcessDescriptions>
```
3. write and handle response
    1.  I've not got to exploring this yet, I'm probably wanting it to open a returned link in new page

I'm wondering if I'm missing any config options that might make this all a bit easier? ability to define the WPS input via configuration rather than the xml?
Something completely different?

Thanks!

**Comment by hardreddata** [comment]
I looked at doing something along these lines while ago and found PyWPS and the GeoServer WPS had some high barriers to entry. We spoke about it here https://gitter.im/TerriaJS/terriajs?at=5f55b39f49a1df0a12d28ee6

I hijacked the WCS functionality to allow me to draw a box on vector point layers and return a .csv (ie: crude vector clip & ship rather than raster). You could feed your box into your processing.

Outside of that gitter link flask and the example from https://www.youtube.com/watch?v=CweFPVewYS0 was simple enough to set up and might be worth a look.

**Reply by stevenjh** [reply]
Nice work digging out the old gitter references, the other links in that thread look helpful!
And see I'm following my own old approach from there of extend terriajs-server.
